Pipeline Data Corruption

EasterTheBunny opened the pull request:
The unit test `TestDivide_Example` was acting flakey in the CI pipeline which suggested a flaw in the divide and
multiply operations. When running the test, the expected result would be one of the input values or the division
result in failure cases. This implied that results were either received out of order or were being sorted incorrectly.

The pipeline runner does a final sort on the results, so that ruled out the received out of order possibility. On
inspection of the sorting index on each task, every index was the zero value. This resulted in occasional correct
and incorrect sorting, causing the test flake.

To correct the problem, the `unset` value for tasks was set to `-1` and new checks were added to enforce 0-based indexing on explicitly set indexes.

samsondav reviewed (changes requested):
Can we add some checks to this to prevent being able to parse a pipeline with ambiguous ordering?

EasterTheBunny committed:
Pipeline Data Corruption

The unit test `TestDivide_Example` was acting flakey in the CI pipeline which suggested a flaw in the divide and
multiply operations. When running the test, the expected result would be one of the input values or the division
result in failure cases. This implied that results were either received out of order or were being sorted incorrectly.

The pipeline runner does a final sort on the results, so that ruled out the received out of order possibility. On
inspection of the sorting index on each task, every index was the zero value. This resulted in occasional correct
and incorrect sorting, causing the test flake.

To correct the problem, the test was updated such that the expected result has an index of `1`, leaving all
other tasks with a `0` index.

EasterTheBunny committed:
fix test

samsondav reviewed (dismissed):
Technically a breaking change and should be marked as such, but since the behaviour if you mixed and matched unspecified+0 index was undefined anyway, it shouldn't affect anything

EasterTheBunny committed:
updated changeset
